google/cloud/internal/debug_string.h

```diff
@@ -15,15 +15,65 @@
 #ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_DEBUG_STRING_H
 #define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_DEBUG_STRING_H
 
+#include "google/cloud/internal/absl_str_cat_quiet.h"
 #include "google/cloud/tracing_options.h"
 #include "google/cloud/version.h"
+#include "absl/strings/string_view.h"
 #include <string>
 
 namespace google {
 namespace cloud {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace internal {
 
+/**
+ * Build strings for use in request/response logging.
+ *
+ * The intent is to produce strings with a format similar to those returned
+ * by `DebugString(google::protobuf::Message const&, TracingOptions const&)`
+ * from `google/cloud/internal/debug_string_protobuf.h` for a proto message.
+ */
+class DebugFormatter {
+ public:
+  DebugFormatter(TracingOptions options, absl::string_view message_name);
+
+  DebugFormatter& SubMessage(absl::string_view message_name);
+  DebugFormatter& EndMessage();
+
+  DebugFormatter& Field(absl::string_view field_name, bool value);
+
+  template <typename T>
+  DebugFormatter& Field(absl::string_view field_name, T const& value) {
+    absl::StrAppend(&str_, Sep(), field_name, ": ", value);
+    return *this;
+  }
+
+  template <typename T>
+  DebugFormatter& QuotedField(absl::string_view field_name, T const& value) {
+    absl::StrAppend(&str_, Sep(), field_name, ": ", "\"", value, "\"");
+    return *this;
+  }
+
+  template <typename T>
+  DebugFormatter& QuotedField(T const& value) {
+    absl::StrAppend(&str_, Sep(), "\"", value, "\"");
+    return *this;
+  }
+
+  DebugFormatter& StringField(absl::string_view field_name, std::string value);
```

Comment:
This member function name is slightly odd, but I am not sure I have better ideas.

Reply:
I don't have any better ideas at the moment either.  Given that it is supposed to differentiate fields for which `truncate_string_field_longer_than` should apply, this seemed like a good first approximation.